### ISSUE TITLE
Error adding tags with WebDAV for non-favorited files

### ISSUE THREAD

**Issue by olivierlemasle** [owncloud/core#16053]
It seems that tags can be added to a file using WebDAV only if the file is already flagged as "favorite".

### Steps to reproduce
1. Create files "test1.txt" and "test2.txt" in the web UI

2. Using WebDAV, set "test1.txt" as "favorite", then add two tags to the file:

1st request:

PROPPATCH http://owncloud-host/owncloud/remote.php/webdav/test1.txt
```xml
<?xml version="1.0"?>
<d:propertyupdate xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns">
  <d:set>
    <d:prop>
      <oc:favorite>1</oc:favorite>
    </d:prop>
  </d:set>
</d:propertyupdate>
```

Response: 

```xml
<?xml version="1.0" encoding="utf-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:oc="http://owncloud.org/ns">
  <d:response>
    <d:href>/owncloud/remote.php/webdav/test1.txt</d:href>
    <d:propstat>
      <d:prop>
        <oc:favorite>1</oc:favorite>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```

2nd request:

PROPPATCH http://owncloud-host/owncloud/remote.php/webdav/test1.txt
```xml
<?xml version="1.0"?>
<d:propertyupdate xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns">
  <d:set>
    <d:prop>
      <oc:tags>
        <oc:tag>tag1</oc:tag>
        <oc:tag>tag2</oc:tag>
      </oc:tags>
    </d:prop>
  </d:set>
</d:propertyupdate>
```

Response:
```xml
<?xml version="1.0" encoding="utf-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:oc="http://owncloud.org/ns">
  <d:response>
    <d:href>/owncloud/remote.php/webdav/test1.txt</d:href>
    <d:propstat>
      <d:prop>
        <oc:tags>
          <oc:tag>tag1</oc:tag>
          <oc:tag>tag2</oc:tag>
        </oc:tags>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```

On the second file, only add the tags:

Request:

PROPPATCH http://owncloud-host/owncloud/remote.php/webdav/test2.txt
```xml
<?xml version="1.0"?>
<d:propertyupdate xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns">
  <d:set>
    <d:prop>
      <oc:tags>
        <oc:tag>tag1</oc:tag>
        <oc:tag>tag2</oc:tag>
      </oc:tags>
    </d:prop>
  </d:set>
</d:propertyupdate>
```

Response:
```xml
<?xml version="1.0" encoding="utf-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:oc="http://owncloud.org/ns">
  <d:response>
    <d:href>/owncloud/remote.php/webdav/test2.txt</d:href>
    <d:propstat>
      <d:prop>
        <oc:tags>
          <oc:tag>tag1</oc:tag>
          <oc:tag>tag2</oc:tag>
        </oc:tags>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```

### Expected behaviour
We should expect to have the two tags on both files.

### Actual behaviour
Only the first file has the tags.

Request:

PROPFIND http://owncloud-host/owncloud/remote.php/webdav/test1.txt
```xml
<?xml version="1.0"?>
<a:propfind xmlns:a="DAV:" xmlns:oc="http://owncloud.org/ns">
  <a:prop><a:getcontenttype/></a:prop>
  <a:prop><oc:tags/></a:prop>
  <a:prop><oc:favorite/></a:prop>
</a:propfind>
```

Response:
```xml
<?xml version="1.0" encoding="utf-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:oc="http://owncloud.org/ns">
  <d:response>
    <d:href>/owncloud/remote.php/webdav/test1.txt</d:href>
    <d:propstat>
      <d:prop>
        <oc:id>00003457ocwk0l0ouedl</oc:id>
        <oc:permissions>RDNVW</oc:permissions>
        <oc:tags>
          <oc:tag>tag1</oc:tag>
          <oc:tag>tag2</oc:tag>
        </oc:tags>
        <oc:favorite>1</oc:favorite>
        <d:getcontenttype>text/plain</d:getcontenttype>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```

Request on second file:

PROPFIND http://owncloud-host/owncloud/remote.php/webdav/test2.txt
```xml
<?xml version="1.0"?>
<a:propfind xmlns:a="DAV:" xmlns:oc="http://owncloud.org/ns">
  <a:prop><a:getcontenttype/></a:prop>
  <a:prop><oc:tags/></a:prop>
  <a:prop><oc:favorite/></a:prop>
</a:propfind>
```

Response:
```xml
<?xml version="1.0" encoding="utf-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:oc="http://owncloud.org/ns">
  <d:response>
    <d:href>/owncloud/remote.php/webdav/test2.txt</d:href>
    <d:propstat>
      <d:prop>
        <oc:id>00003458ocwk0l0ouedl</oc:id>
        <oc:permissions>RDNVW</oc:permissions>
        <oc:tags/>
        <oc:favorite></oc:favorite>
        <d:getcontenttype>text/plain</d:getcontenttype>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```

### Server configuration
**Operating system**: Debian GNU/Linux 7

**Database:**  MySQL 5.5.43-0+deb7u1

**PHP version:** PHP 5.4.39-0+deb7u2

**ownCloud version:** 8.0.3.4

**Updated from an older ownCloud or fresh install:** updated from 8.0.2

#### ownCloud log (data/owncloud.log)
```
{"reqId":"f09499de65baf46c57edd403aa6627ea","remoteAddr":"10.0.8.6","app":"core","message":"OC\\Tags::__construct, tags: Array\n(\n    [0] => OC\\Tagging\\Tag Object\n        (\n            [owner:protected] => admin\n            [type:protected] => files\n            [name:protected] => _$!<Favorite>!$_\n            [id] => 1\n            [_updatedFields:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                )\n\n            [_fieldTypes:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                    [id] => integer\n                )\n\n        )\n\n    [1] => OC\\Tagging\\Tag Object\n        (\n            [owner:protected] => admin\n            [type:protected] => files\n            [name:protected] => amysta\n            [id] => 4\n            [_updatedFields:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                )\n\n            [_fieldTypes:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                    [id] => integer\n                )\n\n        )\n\n    [2] => OC\\Tagging\\Tag Object\n        (\n            [owner:protected] => admin\n            [type:protected] => files\n            [name:protected] => tag1\n            [id] => 2\n            [_updatedFields:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                )\n\n            [_fieldTypes:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                    [id] => integer\n                )\n\n        )\n\n    [3] => OC\\Tagging\\Tag Object\n        (\n            [owner:protected] => admin\n            [type:protected] => files\n            [name:protected] => tag2\n            [id] => 3\n            [_updatedFields:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                )\n\n            [_fieldTypes:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                    [id] => integer\n                )\n\n        )\n\n)\n","level":0,"time":"2015-05-04T16:57:55+00:00","method":"PROPPATCH","url":"\/owncloud\/remote.php\/webdav\/test2.txt"}
{"reqId":"f09499de65baf46c57edd403aa6627ea","remoteAddr":"10.0.8.6","app":"PHP","message":"array_diff(): Argument #2 is not an array at \/var\/www\/owncloud\/lib\/private\/connector\/sabre\/tagsplugin.php#168","level":3,"time":"2015-05-04T16:57:55+00:00","method":"PROPPATCH","url":"\/owncloud\/remote.php\/webdav\/test2.txt"}
{"reqId":"f09499de65baf46c57edd403aa6627ea","remoteAddr":"10.0.8.6","app":"PHP","message":"Invalid argument supplied for foreach() at \/var\/www\/owncloud\/lib\/private\/connector\/sabre\/tagsplugin.php#169","level":3,"time":"2015-05-04T16:57:55+00:00","method":"PROPPATCH","url":"\/owncloud\/remote.php\/webdav\/test2.txt"}
{"reqId":"f09499de65baf46c57edd403aa6627ea","remoteAddr":"10.0.8.6","app":"PHP","message":"array_diff(): Argument #1 is not an array at \/var\/www\/owncloud\/lib\/private\/connector\/sabre\/tagsplugin.php#175","level":3,"time":"2015-05-04T16:57:55+00:00","method":"PROPPATCH","url":"\/owncloud\/remote.php\/webdav\/test2.txt"}
{"reqId":"f09499de65baf46c57edd403aa6627ea","remoteAddr":"10.0.8.6","app":"PHP","message":"Invalid argument supplied for foreach() at \/var\/www\/owncloud\/lib\/private\/connector\/sabre\/tagsplugin.php#176","level":3,"time":"2015-05-04T16:57:55+00:00","method":"PROPPATCH","url":"\/owncloud\/remote.php\/webdav\/test2.txt"}
{"reqId":"351c00b3ad0cabd19bad9ba435444fba","remoteAddr":"10.0.8.6","app":"core","message":"OC\\Tags::__construct, tags: Array\n(\n    [0] => OC\\Tagging\\Tag Object\n        (\n            [owner:protected] => admin\n            [type:protected] => files\n            [name:protected] => _$!<Favorite>!$_\n            [id] => 1\n            [_updatedFields:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                )\n\n            [_fieldTypes:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                    [id] => integer\n                )\n\n        )\n\n    [1] => OC\\Tagging\\Tag Object\n        (\n            [owner:protected] => admin\n            [type:protected] => files\n            [name:protected] => amysta\n            [id] => 4\n            [_updatedFields:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                )\n\n            [_fieldTypes:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                    [id] => integer\n                )\n\n        )\n\n    [2] => OC\\Tagging\\Tag Object\n        (\n            [owner:protected] => admin\n            [type:protected] => files\n            [name:protected] => tag1\n            [id] => 2\n            [_updatedFields:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                )\n\n            [_fieldTypes:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                    [id] => integer\n                )\n\n        )\n\n    [3] => OC\\Tagging\\Tag Object\n        (\n            [owner:protected] => admin\n            [type:protected] => files\n            [name:protected] => tag2\n            [id] => 3\n            [_updatedFields:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                )\n\n            [_fieldTypes:OCP\\AppFramework\\Db\\Entity:private] => Array\n                (\n                    [id] => integer\n                )\n\n        )\n\n)\n","level":0,"time":"2015-05-04T16:59:08+00:00","method":"PROPFIND","url":"\/owncloud\/remote.php\/webdav\/test2.txt"}

```

**Comment by karlitschek** [comment]
@PVince81 Is this a bug?

**Comment by olivierlemasle** [comment]
@karlitschek @PVince81 It really seems a bug.

Looking at the logs (provided in my first comment), we can see that it regards this line:

https://github.com/owncloud/core/blob/v8.0.3/lib/private/connector/sabre/tagsplugin.php#L168

where `$currentTags` is not an array.

NB: To find the proper WebDAV format to set tags, I looked at https://github.com/owncloud/core/pull/12865#issuecomment-67346274 (@PVince81's comment)

**Comment by PVince81** [comment]
I haven't digged into this yet, but if that's what you observe it is likely to be a bug in the classes related to the tags.

Note that "favorite" is a special kind of tag, so when you set it, it maybe initializes something in the DB which is required by tags too. And maybe that "something" isn't initialized properly when setting tags directly. (just speculating, need to look into the code)

**Comment by PVince81** [comment]
Tried to reproduce this on 8.0.8 and got these errors:
```
{"reqId":"cff0c6c7b87c50cc9c14f11fab6a9865","remoteAddr":"127.0.0.1","app":"PHP","message":"array_diff(): Argument #2 is not an array at \/srv\/www\/htdocs\/owncloud\/lib\/private\/connector\/sabre\/tagsplugin.php#168","level":3,"time":"2015-10-12T11:32:34+00:00","method":"PROPPATCH","url":"\/owncloud\/remote.php\/webdav\/test2.txt"}
{"reqId":"cff0c6c7b87c50cc9c14f11fab6a9865","remoteAddr":"127.0.0.1","app":"PHP","message":"Invalid argument supplied for foreach() at \/srv\/www\/htdocs\/owncloud\/lib\/private\/connector\/sabre\/tagsplugin.php#169","level":3,"time":"2015-10-12T11:32:34+00:00","method":"PROPPATCH","url":"\/owncloud\/remote.php\/webdav\/test2.txt"}
{"reqId":"cff0c6c7b87c50cc9c14f11fab6a9865","remoteAddr":"127.0.0.1","app":"PHP","message":"array_diff(): Argument #1 is not an array at \/srv\/www\/htdocs\/owncloud\/lib\/private\/connector\/sabre\/tagsplugin.php#175","level":3,"time":"2015-10-12T11:32:34+00:00","method":"PROPPATCH","url":"\/owncloud\/remote.php\/webdav\/test2.txt"}
{"reqId":"cff0c6c7b87c50cc9c14f11fab6a9865","remoteAddr":"127.0.0.1","app":"PHP","message":"Invalid argument supplied for foreach() at \/srv\/www\/htdocs\/owncloud\/lib\/private\/connector\/sabre\/tagsplugin.php#176","level":3,"time":"2015-10-12T11:32:34+00:00","method":"PROPPATCH","url":"\/owncloud\/remote.php\/webdav\/test2.txt"}
```

And the tags were not set properly.

**Comment by PVince81** [comment]
This https://github.com/owncloud/core/blob/v8.0.8/lib/private/connector/sabre/tagsplugin.php#L155 should be `return [];` instead of `return null;`.
Then it works.

**Comment by PVince81** [comment]
Looks like it works properly in 8.2, it was fixed differently.

**Comment by PVince81** [comment]
Also fixed in 8.1.3 as part of the SabreDAV update, this is the block that was modified: https://github.com/owncloud/core/pull/14151/files#diff-aa826dde9ddbf6ed6621f1b0cdc524b3R161

**Comment by PVince81** [comment]
Please upgrade to 8.1.3 (or the upcoming 8.1.4RC1) or 8.2.

I don't think this is critical enough for a 8.0 specific fix. CC @karlitschek @DeepDiver1975 @cmonteroluque
